segfault due to race in ebpf initialisation

Apply
```diff
diff --git a/probe/endpoint/ebpf.go b/probe/endpoint/ebpf.go
index ff23b4e2..f9d470b5 100644
--- a/probe/endpoint/ebpf.go
+++ b/probe/endpoint/ebpf.go
@@ -7,6 +7,7 @@ import (
        "strconv"
        "sync"
        "syscall"
+       "time"
 
        log "github.com/Sirupsen/logrus"
        "github.com/weaveworks/common/fs"
@@ -96,6 +97,9 @@ func newEbpfTracker() (eventTracker, error) {
                tracer:          t,
        }
 
+       log.Infof("delaying setting of ebpfTracker")
+       time.Sleep(5 * time.Second)
+
        ebpfTracker = tracker
        return tracker, nil
 }
```
and then start a probe
```
$ sudo prog/scope --mode=probe --weave=false
INFO[0000] publishing to: 127.0.0.1:4040                
<probe> INFO: 2017/07/06 14:06:01.763599 command line args: --mode=probe --weave=false
<probe> INFO: 2017/07/06 14:06:01.763676 probe starting, version 0dbea45e, ID 3f9c8145f01926a1
<probe> ERRO: 2017/07/06 14:06:01.766805 Error fetching app details: Get http://127.0.0.1:4040/api: dial tcp 127.0.0.1:4040: getsockopt: connection refused
<probe> INFO: 2017/07/06 14:06:02.267125 delaying setting of ebpfTracker
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x176e70a]

goroutine 73 [running]:
github.com/weaveworks/scope/probe/endpoint.(*EbpfTracker).handleConnection(0x0, 0xc400000003, 0xc4219699f0, 0xc, 0xc421969a10, 0xd, 0x1bba04a, 0x5315, 0xc421969a30, 0xa)
	/go/src/github.com/weaveworks/scope/probe/endpoint/ebpf.go:208 +0x3a
github.com/weaveworks/scope/probe/endpoint.tcpEventCbV4(0x94fc54310ad1, 0x3, 0x531500000003, 0xc421969978, 0x5, 0xc421969990, 0x10, 0x10, 0xc4219699b0, 0x10, ...)
	/go/src/github.com/weaveworks/scope/probe/endpoint/ebpf.go:126 +0x217
github.com/weaveworks/scope/vendor/github.com/weaveworks/tcptracer-bpf/pkg/tracer.NewTracer.func1(0xc422e689c0, 0xc420468240, 0x287bf60, 0xc420468300, 0x287bf58)
	/go/src/github.com/weaveworks/scope/vendor/github.com/weaveworks/tcptracer-bpf/pkg/tracer/tracer.go:84 +0x1d6
created by github.com/weaveworks/scope/vendor/github.com/weaveworks/tcptracer-bpf/pkg/tracer.NewTracer
	/go/src/github.com/weaveworks/scope/vendor/github.com/weaveworks/tcptracer-bpf/pkg/tracer/tracer.go:89 +0x6d3
```

That's because the tracer invokes callback functions that dispatch via a global(!) var which we can only set after initialising the tracer. Chicken and egg. Easy to fix.

--- a/probe/endpoint/ebpf.go
+++ b/probe/endpoint/ebpf.go
@@ -33,8 +33,6 @@ type eventTracker interface {
 	stop()
 }
 
-var ebpfTracker *EbpfTracker
-
 // EbpfTracker contains the sets of open and closed TCP connections.
 // Closed connections are kept in the `closedConnections` slice for one iteration of `walkConnections`.
 type EbpfTracker struct {
@@ -86,54 +84,52 @@ func newEbpfTracker() (eventTracker, error) {
 		return nil, fmt.Errorf("kernel not supported: %v", err)
 	}
 
-	t, err := tracer.NewTracer(tcpEventCbV4, tcpEventCbV6, lostCb)
+	tracker := &EbpfTracker{
+		openConnections: map[string]ebpfConnection{},
+	}
+
+	tracer, err := tracer.NewTracer(tracker.tcpEventCbV4, tracker.tcpEventCbV6, tracker.lostCb)
 	if err != nil {
 		return nil, err
 	}
 
-	tracker := &EbpfTracker{
-		openConnections: map[string]ebpfConnection{},
-		tracer:          t,
-	}
-
-	ebpfTracker = tracker
-
-	t.Start()
+	tracker.tracer = tracer
+	tracer.Start()
 
 	return tracker, nil
 }
 
 var lastTimestampV4 uint64
 
-func tcpEventCbV4(e tracer.TcpV4) {
+func (t *EbpfTracker) tcpEventCbV4(e tracer.TcpV4) {
 	if lastTimestampV4 > e.Timestamp {
 		// A kernel bug can cause the timestamps to be wrong (e.g. on Ubuntu with Linux 4.4.0-47.68)
 		// Upgrading the kernel will fix the problem. For further info see:
 		// https://github.com/iovisor/bcc/issues/790#issuecomment-263704235
 		// https://github.com/weaveworks/scope/issues/2334
 		log.Errorf("tcp tracer received event with timestamp %v even though the last timestamp was %v. Stopping the eBPF tracker.", e.Timestamp, lastTimestampV4)
-		ebpfTracker.dead = true
-		ebpfTracker.stop()
+		t.dead = true
+		t.stop()
 	}
 
 	lastTimestampV4 = e.Timestamp
 
 	if e.Type == tracer.EventFdInstall {
-		ebpfTracker.handleFdInstall(e.Type, int(e.Pid), int(e.Fd))
+		t.handleFdInstall(e.Type, int(e.Pid), int(e.Fd))
 	} else {
 		tuple := fourTuple{e.SAddr.String(), e.DAddr.String(), e.SPort, e.DPort}
-		ebpfTracker.handleConnection(e.Type, tuple, int(e.Pid), strconv.Itoa(int(e.NetNS)))
+		t.handleConnection(e.Type, tuple, int(e.Pid), strconv.Itoa(int(e.NetNS)))
 	}
 }
 
-func tcpEventCbV6(e tracer.TcpV6) {
+func (t *EbpfTracker) tcpEventCbV6(e tracer.TcpV6) {
 	// TODO: IPv6 not supported in Scope
 }
 
-func lostCb(count uint64) {
+func (t *EbpfTracker) lostCb(count uint64) {
 	log.Errorf("tcp tracer lost %d events. Stopping the eBPF tracker", count)
-	ebpfTracker.dead = true
-	ebpfTracker.stop()
+	t.dead = true
+	t.stop()
 }
 
 func tupleFromPidFd(pid int, fd int) (tuple fourTuple, netns string, ok bool) {

--- a/probe/endpoint/ebpf.go
+++ b/probe/endpoint/ebpf.go
@@ -97,6 +97,9 @@ func newEbpfTracker() (eventTracker, error) {
 	}
 
 	ebpfTracker = tracker
+
+	t.Start()
+
 	return tracker, nil
 }
 

--- a/probe/endpoint/ebpf_test.go
+++ b/probe/endpoint/ebpf_test.go
@@ -209,13 +209,12 @@ func TestInvalidTimeStampDead(t *testing.T) {
 		dead:            false,
 		openConnections: map[string]ebpfConnection{},
 	}
-	ebpfTracker = mockEbpfTracker
 	event.Timestamp = 0
-	tcpEventCbV4(event)
+	mockEbpfTracker.tcpEventCbV4(event)
 	event2 := event
 	event2.SPort = 1
 	event2.Timestamp = 2
-	tcpEventCbV4(event2)
+	mockEbpfTracker.tcpEventCbV4(event2)
 	mockEbpfTracker.walkConnections(func(e ebpfConnection) {
 		cnt++
 	})
@@ -227,7 +226,7 @@ func TestInvalidTimeStampDead(t *testing.T) {
 	}
 	cnt = 0
 	event.Timestamp = 1
-	tcpEventCbV4(event)
+	mockEbpfTracker.tcpEventCbV4(event)
 	mockEbpfTracker.walkConnections(func(e ebpfConnection) {
 		cnt++
 	})

